### DISCUSSION TITLE
Best way to handle V2 network messages?

### DISCUSSION THREAD

**Discussion by nyonson** [rust-bitcoin/rust-bitcoin#2959] [Ideas]
BIP324 introduced an encrypted transport protocol, but for reasons listed in the BIP, it also introduced [a new over the wire structure for application messages](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#user-content-v2_Bitcoin_P2P_message_structure).

Over in our BIP324 library we have a [serialization module that handles this](https://github.com/rust-bitcoin/bip324/blob/main/protocol/src/serde.rs), but it is pretty hairy requiring some code to be copy/pasted out of rust-bitcoin. There is some awkwardness between rust-bitcoin's existing `NetworkMessage` and `RawNetworkMessage` types and the new V2 logic since it crosses across both types. I think ideally rust-bitcoin handles the V2 logic, although curious if others think otherwise. In any case, it's not immediately clear to me the best spot to splice it in and if a larger refactor is necessary.

**Comment by apoelstra** [comment]
This makes sense to me.

In rust-bitcoin, the p2p module has never seen a lot of love. Initially I hacked together half a dozen messages and then people grafted on other messages, but it's never been complete and never been modernized beyond the work required to keep clippy happy. During our crate-smashing project we will split it out into its own crate `bitcoin-p2p` or something like that, and during that process I expect we'll investigate the API more closely and do some changes.

In the meantime if you want to help move that forward, I would welcome any contributions to the module! Generalizing it to support V2 sounds reasonable to me; at the very least we should do better with them than to just have parsing errors or whatever.

**Reply by Kixunil** [reply]
I have no clue what the differences are but it sounds to me that we'll just want different structs for them, so I'd like contributions to go that direction unless they can provide a reason explaining why it's not good enough.

**Reply by nyonson** [reply]
I wanted to know the story/plan of the network message types before doing too much, but sounds like they have just grown organically so I'll dive in and see what some options are for supporting V2.
